File: billsplitter.py
import random


class SplitBillCalculator:
    def __init__(self):
        self.friends = {}
        self.lucky_feature = False
        self.lucky = None

    def get_number_of_friends(self):
        return int(input('Enter the number of friends joining (including you):\n'))

    def get_friend_names(self, num_friends):
        print('Enter the name of every friend (including you), each on a new line:')
        for _ in range(num_friends):
            friend = input()
            self.friends[friend] = 0

    def get_total_bill(self):
        return int(input('Enter the total bill value:\n'))

    def use_lucky_feature(self):
        answer = input('Do you want to use the "Who is lucky?" feature? Write Yes/No:\n')
        if answer.lower() == 'yes':
            self.lucky = random.choice(list(self.friends.keys()))
            print(f'{self.lucky} is the lucky one!')
            self.lucky_feature = True
        else:
            print('No one is going to be lucky')

    def calculate_and_print_split_bill(self, bill):
        if self.lucky_feature:
            price = round(bill / (len(self.friends) - 1), 2)
        else:
            price = round(bill / len(self.friends), 2)

        for friend in self.friends:
            if friend != self.lucky:
                self.friends[friend] += price

        print(self.friends)

    def main(self):
        num_friends = self.get_number_of_friends()

        if num_friends > 0:
            self.get_friend_names(num_friends)
            total_bill = self.get_total_bill()
            self.use_lucky_feature()
            self.calculate_and_print_split_bill(total_bill)
        else:
            print('No one is joining for the party')


if __name__ == "__main__":
    split_bill_calculator = SplitBillCalculator()
    split_bill_calculator.main()

File: billsplitter.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

func main() {
	rand.NewSource(time.Now().UnixNano())

	friends := make(map[string]float64)
	luckyFeature := false
	var lucky string

	fmt.Println("Enter the number of friends joining (including you):")
	var num int
	fmt.Scan(&num)
	fmt.Println()

	if num > 0 {
		fmt.Println("Enter the name of every friend (including you), each on a new line:")
		for i := 0; i < num; i++ {
			var friend string
			fmt.Scan(&friend)
			friends[friend] = 0
		}
		fmt.Println()
		fmt.Println("Enter the total bill value:")
		var bill float64
		fmt.Scan(&bill)
		fmt.Println()
		fmt.Println("Do you want to use the \"Who is lucky?\" feature? Write Yes/No:")
		var answer string
		fmt.Scan(&answer)
		if answer == "Yes" || answer == "yes" {
			fmt.Println()
			lucky = getRandomFriend(friends)
			fmt.Printf("%s is the lucky one!\n", lucky)
			luckyFeature = true
		} else {
			fmt.Println()
			fmt.Println("No one is going to be lucky")
		}
		var price float64
		if luckyFeature {
			price = round(bill/float64(len(friends)-1), 2)
		} else {
			price = round(bill/float64(len(friends)), 2)
		}
		fmt.Println()
		for friend, debt := range friends {
			if friend != lucky {
				friends[friend] = debt + price
			}
		}
		fmt.Println(friends)
	} else {
		fmt.Println("No one is joining for the party")
	}
}

func round(num float64, decimalPlaces int) float64 {
	precision := math.Pow(10, float64(decimalPlaces))
	return math.Round(num*precision) / precision
}

func getRandomFriend(friends map[string]float64) string {
	keys := make([]string, 0, len(friends))
	for key := range friends {
		keys = append(keys, key)
	}
	return keys[rand.Intn(len(keys))]
}
